File: algorithms/java-quiz/src/main/java/com/github/kimhyunjin/inflearn/stackqueue/Postfix.java
package com.github.kimhyunjin.inflearn.stackqueue;

import java.util.Scanner;
import java.util.Stack;

public class Postfix {

    private static int solution(String exp) {
        Stack<Integer> stack = new Stack<>();
        for (char c : exp.toCharArray()) {
            if (Character.isDigit(c)) {
                stack.push(Character.getNumericValue(c));
            } else {
                int a = stack.pop();
                int b = stack.pop();

                if (c == '+') {
                    stack.push(b + a);
                } else if (c == '-') {
                    stack.push(b - a);
                } else if (c == '*') {
                    stack.push(b * a);
                } else {
                    stack.push(b / a);
                }
            }
        }
        return stack.pop();
    }

    public static void main(String[] args){
        Scanner in = new Scanner(System.in);
        String exp = in.next();
        System.out.println(solution(exp));
        return ;
    }
}

File: content/algorithms/python-quiz/inflearn/stackqueue/postfix.py
def solution(str):
    strList = list(str)
    stack = []
    for s in strList:
        if(s.isdigit()):
            stack.append(int(s))
        else:
            a = stack.pop(-1)
            b = stack.pop(-1)
            if (s == '+'): stack.append(b + a)
            elif (s == '-'): stack.append(b - a)
            elif (s == '*'): stack.append(b * a)
            elif (s == '/'): stack.append(b / a)
    return stack[-1]

def test():
    given = "352+*9-"
    expect = 12
    actual = solution(given)
    if (expect == actual): print(True)
    else: print(False)

test()
        
